solr/core/src/java/org/apache/solr/request/SolrRequestInfo.java

```diff
@@ -95,25 +99,33 @@ public static void clearRequestInfo() {
    * we expect it to be empty by now because all "set" calls need to be balanced with a "clear".
    */
   public static void reset() {
+    log.trace("reset()");
     Deque<SolrRequestInfo> stack = threadLocal.get();
-    boolean isEmpty = stack.isEmpty();
+    assert stack.isEmpty() : "SolrRequestInfo Stack should have been cleared.";
     while (!stack.isEmpty()) {
       SolrRequestInfo info = stack.pop();
-      closeHooks(info);
+      info.close();
     }
-    assert isEmpty : "SolrRequestInfo Stack should have been cleared.";
   }
 
-  private static void closeHooks(SolrRequestInfo info) {
-    if (info.closeHooks != null) {
-      for (Closeable hook : info.closeHooks) {
+  private synchronized void close() {
```

Comment:
does the whole method need to be sync?

Reply:
Reducing synchronized scope to condition seems working. Is it worth to push into PRs' branch.

Reply:
I think it should be the whole method -- for the refCount & closeHook manipulation.  Besides; it shouldn't be contented.